server/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java

```diff
@@ -214,20 +216,44 @@ public String toString() {
         private final String restoreUUID;
         private final Snapshot snapshot;
         private final String index;
+
+        @Nullable
+        private final IndexId indexId;
         private final Version version;
 
-        public SnapshotRecoverySource(String restoreUUID, Snapshot snapshot, Version version, String index) {
+        public SnapshotRecoverySource(String restoreUUID, Snapshot snapshot, Version version, IndexId indexId) {
             this.restoreUUID = restoreUUID;
             this.snapshot = Objects.requireNonNull(snapshot);
             this.version = Objects.requireNonNull(version);
-            this.index = Objects.requireNonNull(index);
+            this.indexId = Objects.requireNonNull(indexId);
+            this.index = indexId.getName();
         }
 
         SnapshotRecoverySource(StreamInput in) throws IOException {
             restoreUUID = in.readString();
             snapshot = new Snapshot(in);
             version = Version.readVersion(in);
             index = in.readString();
+            if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
```

Comment:
the approach here does not allow us to later switch to just reading directly IndexId from stream after backport to 7.x.

Can you change things so that you conditionally do `in.readString();` or `new IndexId(in)`

Reply:
Tell me if I'm missing something here. But doesn't this break in the edge case of:

1. Running mixed cluster with old version master
2. Fail over to new version master
3. New version master sends the existing recovery source over the wire because it still has `null` for the uuid because the old master didn't add that when creating the `RecoverySource`? (am I missing a spot where this is upgraded/recreated during master fail-over?

Reply:
Maybe we should differentiate the `in.readString()` / `new IndexId(in)` serialization part and make the logic only rely on `IndexId`, both on master and 7.x, and INDEX_UUID_NA_VALUE to resolve the index id later when needed?